### PR TITLE
run rules with type-checker

### PR DESCRIPTION
**Pull request by vilchik-elena**
Fixes SonarSource/SonarTS#105

### CODE REVIEW

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java [hunk 1]
```diff
@@ -25,69 +25,91 @@
...
 import org.apache.commons.io.IOUtils;
+import org.sonar.api.batch.BatchSide;
 import org.sonar.api.batch.ScannerSide;
+import org.sonar.api.config.Settings;
 import org.sonar.api.utils.command.Command;
 import org.sonar.api.utils.log.Logger;
 import org.sonar.api.utils.log.Loggers;
 import org.sonarsource.api.sonarlint.SonarLintSide;
 
+@BatchSide
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java]: MINOR Remove this use of "BatchSide"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java [hunk 2]
```diff
@@ -47,49 +45,71 @@
...
-  SonarTSCoreBundle(String bundleFile, String... entryPoints) {
-    this.coreFile = bundleFile;
-    this.entryPoints = entryPoints;
-  }
+  // location inside sonarts-core bundle
+  private static final String TSLINT_LOCATION = "/node_modules/tslint/bin/tslint";
+  private static final String SONAR_LOCATION = "/bin/sonar";
 
+  /**
+   * Extracting "sonarts-core.zip" (containing typescript, tslint and tslint-sonarts)
```
Comment (pynicolas) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java]: That's a useful comment which helps the review. Thanks!

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java [hunk 3]
```diff
@@ -47,49 +45,71 @@
...
-    String runnerFolder = deployDestination + "/sonarts-core";
-    // TODO consider setting permissions here on the fly instead of on-deploy
-    // TODO consider not using the tslint script but rather calling node directly
-    // node -e 'require("runnerFolder + /node_modules/tslint/lib/tslint-cli")' -- --config  .....
-    Command command = Command.create(runnerFolder + "/node_modules/tslint/bin/tslint");
-    command.addArgument("--config").addArgument(runnerFolder + "/tslint.json");
+  public Command createRuleCheckCommand(File projectBaseDir, File deployDestination, Settings settings) {
+    File sonartsCoreDir = new File(deployDestination, "sonarts-core");
+    File tslintExecutable = new File(sonartsCoreDir, TSLINT_LOCATION);
+    setExecutablePermissions(tslintExecutable);
```
Comment (pynicolas) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java]: It seems strange to me to have moved that line here: `createRuleCheckCommand` should not have side effects unlike `deploy`.

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java [hunk 4]
```diff
@@ -47,49 +45,71 @@
...
+  /**
+   * Builds command to run "sonar", which is making side information calculation (metrics, highlighting etc.)
+   */
   @Override
-  public Command createSonarCommand(String projectSourcesRoot, String deployDestination) {
-    return Command.create(deployDestination + "/sonarts-core/bin/sonar");
+  public Command createSonarCommand(File deployDestination) {
+    File sonartsCoreDir = new File(deployDestination, "sonarts-core");
+    File sonarExecutable = new File(sonartsCoreDir, SONAR_LOCATION);
+    setExecutablePermissions(sonarExecutable);
```
Comment (pynicolas) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java]: Same thing as above.

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java [hunk 5]
```diff
@@ -116,20 +136,10 @@ private static void extract(File copiedFile) throws IOException {
...
-          permissions.add(PosixFilePermission.GROUP_EXECUTE);
-          permissions.add(PosixFilePermission.OTHERS_EXECUTE);
-          Files.setPosixFilePermissions(entryPoint, permissions);
-        } catch (IOException e) {
-          LOG.error("Failed to set permissions for file " + entryPoint.toString(), e);
-        }
-      });
+  private static void setExecutablePermissions(File target) {
+    if (!target.setExecutable(true, false)) {
+      // (Lena) I think we should fail instead
```
Comment (pynicolas) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java]: Probably. The javadoc for `setExecutable` may help: 

> If executable is false and the underlying file system does not implement an execute permission, then the operation will fail.

Comment (vilchik-elena) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SonarTSCoreBundle.java]: @pynicolas I will make it fail when unit test improved. Currently if I replace `LOG.error` with `throw` my unit test will fail

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundle.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+import org.apache.commons.io.IOUtils;
+import org.sonar.api.batch.BatchSide;
+import org.sonar.api.batch.ScannerSide;
+import org.sonar.api.config.Settings;
+import org.sonar.api.utils.command.Command;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonarsource.api.sonarlint.SonarLintSide;
+
+@BatchSide
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundle.java]: MINOR Remove this use of "BatchSide"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundle.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+@ScannerSide
+@SonarLintSide
+public class SonarTSCoreBundle implements ExecutableBundle {
+
+  private static final Logger LOG = Loggers.get(SonarTSCoreBundle.class);
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundle.java]: MAJOR Remove this unused "LOG" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 8]
```diff
@@ -73,35 +77,38 @@ public void describe(SensorDescriptor sensorDescriptor) {
...
     LOG.info("Rules execution");
-    Failure[] failures = runRules(deployDestination, projectSourcesRoot);
+    Failure[] failures = runRules(executableBundle, projectBaseDir, sensorContext.settings());
     saveFailures(sensorContext, failures);
   }
 
-  private Failure[] runRules(String deployDestination, String projectSourcesRoot) {
-    Command command = coreBundle.createRuleCheckCommand(projectSourcesRoot, deployDestination);
-    String rulesOutput = executeExternalScript(command);
+  private Failure[] runRules(ExecutableBundle executableBundle, File projectBaseDir, Settings settings) {
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: MAJOR Make "runRules" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 9]
```diff
@@ -160,21 +166,26 @@ private static void saveMetric(SensorContext sensorContext, InputFile inputFile,
...
     }
   }
 
-  @VisibleForTesting
-  protected void saveFailures(SensorContext sensorContext, Failure[] failures) {
+  private static String failedMessage(String command) {
+    return String.format("Failed to run external process `%s`", command);
+  }
+
+  private void saveFailures(SensorContext sensorContext, Failure[] failures) {
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: MAJOR Make "saveFailures" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 10]
```diff
@@ -192,8 +203,7 @@ protected void saveFailures(SensorContext sensorContext, Failure[] failures) {
     }
   }
 
-  @VisibleForTesting
-  protected void saveHighlights(SensorContext sensorContext, Highlight[] highlights, InputFile inputFile) {
+  private void saveHighlights(SensorContext sensorContext, Highlight[] highlights, InputFile inputFile) {
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: MAJOR Make "saveHighlights" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [executable](https://github.com/SonarSource/SonarTS-Plugin/blob/a7576d99bec5f00977ca84480870ce76bf266c9d/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable): Add a 'package-info.java' file to document the 'executable' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundle.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.sonar.api.config.Settings;
+import org.sonar.api.utils.command.Command;
+
+public class SonarTSCoreBundle implements ExecutableBundle {
+
+  // relative location inside sonarts-core bundle
+  private static final String TSLINT_LOCATION = "node_modules/tslint/bin/tslint";
+  private static final String SONAR_LOCATION = "bin/sonar";
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundle.java]: MAJOR Remove this unused "LOG" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

File: sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundleFactory.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugin.typescript.executable;
+
+import java.io.File;
+import org.sonar.api.batch.BatchSide;
+import org.sonar.api.batch.ScannerSide;
+import org.sonarsource.api.sonarlint.SonarLintSide;
+
+@BatchSide
```
Comment (SonarTech) [inline sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable/SonarTSCoreBundleFactory.java]: MINOR Remove this use of "BatchSide"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [executable](https://github.com/SonarSource/SonarTS-Plugin/blob/24cb7fcf38f48bc2ae531d3e1d6f6f1c6c576dfa/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/executable): Add a 'package-info.java' file to document the 'executable' package [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java [hunk 13]
```diff
@@ -54,65 +57,98 @@
...
-    verify(fileLinesContext).save();
-    verifyNoMoreInteractions(fileLinesContext);
+  @Test
+  public void should_fail_when_stdErr_tslint_is_not_empty() throws Exception {
+    thrown.expect(IllegalStateException.class);
+    thrown.expectMessage("Failed to run external process `cat not_existing_file`");
+    // it would be nice to assert thrown.expectCause available since jUnit 4.11
 
-    verify(noSonarFilter).noSonarInFile(eq(inputFile), eq(Sets.newHashSet(24)));
+    TestBundleFactory testBundle = new TestBundleFactory().tslint("cat", "not_existing_file");
```
Comment (pynicolas) [inline sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ExternalTypescriptSensorTest.java]: This line should come before...

File: sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/SonarTSCoreBundleTest.java [hunk 14]
```diff
@@ -30,14 +34,61 @@
...
+  @Test
+  public void should_set_executable_rights() throws Exception {
+    new SonarTSCoreBundleFactory("/testBundle.zip").createAndDeploy(DEPLOY_DESTINATION);
 
-    Command sonarCommand = bundle.createTsMetricsCommand();
-    assertThat(sonarCommand.toCommandLine()).isEqualTo("/deployDestination/sonarts-core/bin/sonar");
+    File tslintExecutable = new File(DEPLOY_DESTINATION, "sonarts-core/node_modules/tslint/bin/tslint");
+    File tsMetricsExecutable = new File(DEPLOY_DESTINATION, "sonarts-core/bin/sonar");
+
+    assertThat(tslintExecutable.canExecute()).isTrue();
```
Comment (pynicolas) [inline sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/SonarTSCoreBundleTest.java]: Are we sure that the files in the zip are not executable? Do we have an assertion for that? If the zip was created in the test, we could.

**Pull request merged**